File: ga/glm4_result_2024_6/result/1968G2_cpp_glm/4_1968G2.cpp
use std::cmp::{Ordering, min};
use std::io::{self, BufRead};

const N: usize = 2e5 as usize + 3;
const MAXN: usize = N + 1;

fn find(p: &mut [usize], mut k: usize) -> usize {
    while k <= N {
        if p[k] == k {
            return k;
        }
        k = p[k];
    }
    panic!("Invalid index");
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();
    let t = lines.next().unwrap().unwrap().parse::<usize>().unwrap();

    for _ in 0..t {
        let mut line = lines.next().unwrap().unwrap();
        let n = line.parse::<usize>().unwrap();
        let x = line.parse::<usize>().unwrap();
        let y = line.parse::<usize>().unwrap();
        let a = line.chars().collect::<Vec<char>>();

        let mut p = vec![0; MAXN];
        let mut t = (1..=n).collect::<Vec<usize>>();
        let mut ans = vec![0; MAXN];
        let mut z = vec![0; MAXN];

        z[1] = n;

        for i in 2..=n {
            let mut j = 0;
            while j + z[j] > i {
                z[i] = min(z[i - j + 1], j + z[j] - i);
            }
            while i + z[i] <= n && a[i + z[i]] == a[1 + z[i]] {
                z[i] += 1;
            }
            if i + z[i] > j + z[j] {
                j = i;
            }
        }

        t.sort_by(|&i, &j| z[i].cmp(&z[j]));

        for i in 1..=n {
            let mut k = 1;
            while k <= n && z[t[k]] < i {
                p[t[k]] = t[k] + 1;
                k += 1;
            }
            let mut c = 0;
            for j in 1..=n {
                c += 1;
                j = find(&mut p, j + i);
            }
            ans[c] = i;
        }

        for i in (0..=n).rev() {
            ans[i] = max(ans[i], ans[i + 1]);
        }

        for i in x..=y {
            print!("{} ", ans[i]);
        }
        println!();
    }
}

File: ga/gpt3_5_result_2024_6/result/1968G2_cpp_gpt/4_1968G2.rs
use std::cmp;

const N: usize = 2e5 + 3;

fn find(k: usize, p: &mut [usize]) -> usize {
    if k > p.len() || p[k] == k {
        k
    } else {
        p[k] = find(p[k], p);
        p[k]
    }
}

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let t: usize = input.trim().parse().unwrap();

    for _ in 0..t {
        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut parts = input.split_whitespace();
        let n: usize = parts.next().unwrap().parse().unwrap();
        let x: usize = parts.next().unwrap().parse().unwrap();
        let y: usize = parts.next().unwrap().parse().unwrap();
        let a: String = parts.next().unwrap().parse().unwrap();

        let mut p: Vec<usize> = vec![0; n + 2];
        let mut t: Vec<usize> = vec![0; n + 2];
        let mut ans: Vec<usize> = vec![0; n + 2];
        let mut z: Vec<usize> = vec![0; n + 2];

        z[1] = n;
        let mut j = 0;
        for i in 2..=n {
            if j + z[j] > i {
                z[i] = cmp::min(z[i - j + 1], j + z[j] - i);
            }
            while a.chars().nth(i + z[i] - 1) == a.chars().nth(1 + z[i]) {
                z[i] += 1;
            }
            if i + z[i] > j + z[j] {
                j = i;
            }
        }

        let mut t_sorted: Vec<usize> = (1..=n).collect();
        t_sorted.sort_by(|&i, &j| z[i].cmp(&z[j]));

        let mut k = 1;
        for i in 1..=n {
            let mut c = 0;
            while k <= n && z[t_sorted[k - 1]] < i {
                p[t_sorted[k - 1]] = t_sorted[k - 1] + 1;
                k += 1;
            }
            let mut j = 1;
            while j <= n {
                j = find(j + i, &mut p);
                c += 1;
            }
            ans[c] = i;
        }

        for i in (1..=n).rev() {
            ans[i] = cmp::max(ans[i], ans[i + 1]);
        }

        for i in x..=y {
            print!("{} ", ans[i]);
        }
        println!();
    }
}